<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>svcnvplus_howto â€¢ svcnvplus</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="svcnvplus_howto">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">svcnvplus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/svcnvplus_howto.html">svcnvplus_howto</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>svcnvplus_howto</h1>
            
      
      
      <div class="hidden name"><code>svcnvplus_howto.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><p><code>svncvplus</code> is an R package designed for downstream analyses of DNA copy number variations (CNV) and other structural variants (SV). CNV data can be derived from genotyping and CGH arrays, as well as next generation sequencing; different segmentation algorithms are used to obtain dosage variations (duplications and deletions) across the genome. Alternatively SV calls can be inferred from discordantly aligned reads from whole genome sequencing (WGS) using different algorithms (e.g <a href="https://github.com/Illumina/manta">manta</a>, <a href="https://github.com/arq5x/lumpy-sv">lumpy</a>, etc). SV calls provide linkage information from discordantly aligned reads and read pairs, allowing the discovery of chromosomal translocations and variants that do not necesarily involve dosage change, such as inversions and insertions. Segmentation CNVs and alignment based SV calls produce orthogonal as well as complementary results. The integration of both data types can by highly informative to understand the somatic alterations driving many cancers and is essential to characterize complex chromosomal alterations such as chromothripsis and chromoplexy. In summary, <code>svcnvplus</code> is a toolkit for the integration of SV datasets, visualization and characterization of complex somatic SVs.</p>
<div id="index" class="section level2">
<h2 class="hasAnchor">
<a href="#index" class="anchor"></a>Index:</h2>
<ul>
<li><a href="#install-svcnvplus">Install svcnvplus</a></li>
<li><a href="#input-data">Input data</a></li>
<li>
<a href="#validate-data-types">Validate data types</a>
<ul>
<li><a href="#validate-segmentation-data-format">Validate segmentation data format</a></li>
<li><a href="#validate-structural-variant-data-format">Validate structural variant data format</a></li>
</ul>
</li>
<li>
<a href="#cnv-analysys-and-visualization">CNV analysys and visualization</a>
<ul>
<li><a href="#cnv-frequency-plot">CNV frequency plot</a></li>
<li><a href="#gene-level-cnv">Gene-level CNV</a></li>
<li><a href="#percent-genome-change">Percent genome change</a></li>
</ul>
</li>
<li><a href="#co-localization-of-breakpoints">Co-localization of breakpoints</a></li>
<li><a href="#breakpoint-burden">Breakpoint burden</a></li>
<li>
<a href="#identification-of-shattered-regions">Identification of shattered regions</a>
<ul>
<li><a href="#chromosome-shattering-using-segmentation-data-only">Chromosome shattering using segmentation data only</a></li>
<li><a href="#chromosome-shattering-using-segmentation-and-sv-data">Chromosome shattering using segmentation and SV data</a></li>
</ul>
</li>
<li>
<a href="#recurrently-shattered-regions">Recurrently shattered regions</a>
<ul>
<li><a href="#recurrently-shattered-regions-plot">Recurrently shattered regions plot</a></li>
</ul>
</li>
<li>
<a href="#recurrently-altered-genes">Recurrently altered genes</a>
<ul>
<li><a href="amplifications-and-deep-deletions">Amplifications and deep deletions</a></li>
<li><a href="#recurrently-altered-genes-using-cnv-breakpoints">Recurrently altered genes using CNV breakpoints</a></li>
<li><a href="#recurrently-altered-genes-using-sv-breakpoints">Recurrently altered genes using SV breakpoints</a></li>
</ul>
</li>
</ul>
</div>
<div id="install-svcnvplus" class="section level2">
<h2 class="hasAnchor">
<a href="#install-svcnvplus" class="anchor"></a>Install svcnvplus</h2>
<p>Install development version from GitHub devtools::install_github(gonzolgarcia/svcnvplus")</p>
</div>
<div id="input-data" class="section level2">
<h2 class="hasAnchor">
<a href="#input-data" class="anchor"></a>Input data</h2>
<p>Two data types are allowed: * CNV segmentation data * SV calls</p>
<p>To explore the basic functionalities of svcnvplus, several data.frames are included <code><a href="../reference/segdat_lung_ccle.html">svcnvplus::segdat_lung_ccle</a></code> and <code>svcnvplus::svdat_lung_ccle</code>. This data.frames contain information about CNV segments and structural variants respectively for lung cancer derived cell lines. This data was obtained from <a href="https://depmap.org/portal/download/" class="uri">https://depmap.org/portal/download/</a>. Both data.frames are lazy loaded with <code>svcnvplus</code></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(svcnvplus)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(nbl_segdat)</a></code></pre></div>
<pre><code>## [1] 17680     6</code></pre>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(nbl_segdat)</a></code></pre></div>
<pre><code>##   Sample Chromosome    Start      End Num_markers  Seg_CN
## 1 PAISNS          1    11000   833000         337 -0.0270
## 2 PAISNS          1   835000  2715000         916 -1.0257
## 3 PAISNS          1  2717000  5969000        1552 -0.8316
## 4 PAISNS          1  5971000 12481000        3256 -0.9593
## 5 PAISNS          1 12483000 12777000         148 -0.2305
## 6 PAISNS          1 12779000 15551000        1287 -0.0083</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(nbl_svdat)</a></code></pre></div>
<pre><code>## [1] 7366    8</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(nbl_svdat)</a></code></pre></div>
<pre><code>##   TARGET.USI LeftChr LeftPosition LeftStrand RightChr RightPosition RightStrand
## 1     PAISNS    chr1     12481576          -     chr7     123358964           +
## 2     PAISNS    chr1    120543859          -     chr2      65103235           -
## 3     PAISNS    chr2    231680218          +    chr21      40045998           +
## 4     PAISNS    chr3     54814482          -    chr17      42657036           +
## 5     PAISNS    chr4     97761321          -     chr4      97765146           +
## 6     PAISNS    chr4    190936134          +     chr9      68411170           +
##   Type
## 1  TRA
## 2  TRA
## 3  TRA
## 4  TRA
## 5  INV
## 6  TRA</code></pre>
</div>
<div id="validate-data-types" class="section level2">
<h2 class="hasAnchor">
<a href="#validate-data-types" class="anchor"></a>Validate data types</h2>
<ul>
<li>
<code><a href="../reference/validate.seg.html">validate.seg()</a></code> check format for segmentation data.frame</li>
<li>
<code><a href="../reference/validate.sv.html">validate.sv()</a></code> chrek format of a Structural variant data.frame</li>
</ul>
<div id="validate-segmentation-data-format" class="section level3">
<h3 class="hasAnchor">
<a href="#validate-segmentation-data-format" class="anchor"></a>Validate segmentation data format</h3>
<p>Validate and format segmentation data.frame to be used by svcnvplus package toolkit</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">segdf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/validate.seg.html">validate.seg</a></span>(nbl_segdat)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(segdf)</a></code></pre></div>
<pre><code>##   sample chrom    start      end probes segmean
## 1 PAISNS  chr1    11000   833000    337 -0.0270
## 2 PAISNS  chr1   835000  2715000    916 -1.0257
## 3 PAISNS  chr1  2717000  5969000   1552 -0.8316
## 4 PAISNS  chr1  5971000 12481000   3256 -0.9593
## 5 PAISNS  chr1 12483000 12777000    148 -0.2305
## 6 PAISNS  chr1 12779000 15551000   1287 -0.0083</code></pre>
</div>
<div id="validate-structural-variant-data-format" class="section level3">
<h3 class="hasAnchor">
<a href="#validate-structural-variant-data-format" class="anchor"></a>Validate structural variant data format</h3>
<p>Validate and format structural variant data.frame to be used by svcnvplus package toolkit</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">svdf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/validate.sv.html">validate.sv</a></span>(nbl_svdat)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(svdf)</a></code></pre></div>
<pre><code>##   sample chrom1      pos1 strand1 chrom2      pos2 strand2 svclass
## 1 PAISNS   chr1  12481576       -   chr7 123358964       +     TRA
## 2 PAISNS   chr1 120543859       -   chr2  65103235       -     TRA
## 3 PAISNS   chr2 231680218       +  chr21  40045998       +     TRA
## 4 PAISNS   chr3  54814482       -  chr17  42657036       +     TRA
## 5 PAISNS   chr4  97761321       -   chr4  97765146       +     INV
## 6 PAISNS   chr4 190936134       +   chr9  68411170       +     TRA</code></pre>
</div>
</div>
<div id="cnv-analysys-and-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#cnv-analysys-and-visualization" class="anchor"></a>CNV analysys and visualization</h2>
<div id="cnv-frequency-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#cnv-frequency-plot" class="anchor"></a>CNV frequency plot</h3>
<p>Visualization of CNV frequencies across the genome</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">cnv_freq &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cnv.freq.plot.html">cnv.freq.plot</a></span>(segdf)  <span class="co"># plot cnv frequencies</span></a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="svcnvplus_howto_files/figure-html/plot1-1.png" alt="Genome wide CNV frequencies" width="864"><p class="caption">
Genome wide CNV frequencies
</p>
</div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(cnv_freq<span class="op">$</span>freqsum)  <span class="co"># data.frame contains every genomic bin </span></a></code></pre></div>
<pre><code>##                       chr   start     end freq.gains freq.loss
## chr1_11000_1011000   chr1   11000 1011000 0.02962963 0.3333333
## chr1_1011000_2011000 chr1 1011000 2011000 0.02962963 0.3333333
## chr1_2011000_3011000 chr1 2011000 3011000 0.02222222 0.3333333
## chr1_3011000_4011000 chr1 3011000 4011000 0.02222222 0.3333333
## chr1_4011000_5011000 chr1 4011000 5011000 0.02222222 0.3407407
## chr1_5011000_6011000 chr1 5011000 6011000 0.02222222 0.3407407</code></pre>
</div>
<div id="gene-level-cnv" class="section level3">
<h3 class="hasAnchor">
<a href="#gene-level-cnv" class="anchor"></a>Gene level CNV</h3>
<p>Generate a matrix with gene level CNVs from a segmentation file</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">gene_cnv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/gene.cnv.html">gene.cnv</a></span>(segdf_clean,<span class="dt">genome.v =</span> <span class="st">"hg19"</span>,<span class="dt">fill.gaps =</span> <span class="ot">TRUE</span>,<span class="dt">verbose=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/dim.html">dim</a></span>(gene_cnv<span class="op">$</span>cnvmat)</a></code></pre></div>
<pre><code>## [1] 22886   135</code></pre>
</div>
<div id="percent-genome-change" class="section level3">
<h3 class="hasAnchor">
<a href="#percent-genome-change" class="anchor"></a>Percent genome change</h3>
<p>Returns per sample percentage of genome with CNV</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">pct_change &lt;-<span class="st"> </span><span class="kw"><a href="../reference/pct.genome.changed.html">pct.genome.changed</a></span>(segdf)</a></code></pre></div>
</div>
</div>
<div id="co-localization-of-breakpoints" class="section level2">
<h2 class="hasAnchor">
<a href="#co-localization-of-breakpoints" class="anchor"></a>Co-localization of breakpoints</h2>
<p>CNVs (duplications and deletions) are a subset of all structural variants; Both CNV segmentation profiles and structural variants derived from discordant alignments produce orthogonal calls for these genomic events. The function <code>match.variant.breaks</code> allows identifying common breakpoints called by both approaches.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">common.breaks &lt;-<span class="st"> </span><span class="kw"><a href="../reference/match.variant.breaks.html">match.variant.breaks</a></span>(segdf,svdf,<span class="dt">maxgap=</span><span class="dv">50000</span>,<span class="dt">low.cov =</span> <span class="ot">NULL</span>,<span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"></a>
<a class="sourceLine" id="cb20-3" data-line-number="3">restab &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(common.breaks<span class="op">$</span>restab)[<span class="kw"><a href="https://rdrr.io/r/base/order.html">order</a></span>(common.breaks<span class="op">$</span>restab<span class="op">$</span>total.sv),]</a>
<a class="sourceLine" id="cb20-4" data-line-number="4">m2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sprintf.html">sprintf</a></span>(<span class="st">"%.1f"</span>,<span class="dv">100</span><span class="op">*</span><span class="kw"><a href="https://rdrr.io/r/base/mean.html">mean</a></span>(restab<span class="op">$</span>matched.sv<span class="op">/</span>restab<span class="op">$</span>total.sv))</a>
<a class="sourceLine" id="cb20-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span>(restab<span class="op">$</span>matched.sv,restab<span class="op">$</span>total.sv <span class="op">-</span><span class="st"> </span>restab<span class="op">$</span>matched.sv),</a>
<a class="sourceLine" id="cb20-6" data-line-number="6">        <span class="dt">border=</span><span class="ot">NA</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">xlab=</span><span class="st">""</span>,<span class="dt">horiz=</span><span class="ot">FALSE</span>,<span class="dt">cex.main=</span>.<span class="dv">7</span>,<span class="dt">cex.names=</span>.<span class="dv">4</span>, <span class="dt">names=</span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(restab))</a>
<a class="sourceLine" id="cb20-7" data-line-number="7"><span class="kw"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"top"</span>,<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"SV breaks matched by CNV breaks</span><span class="ch">\n</span><span class="st">"</span>,<span class="st">"Average = "</span>,m2,<span class="st">"%"</span>,<span class="dt">sep=</span><span class="st">""</span>),<span class="dt">bty=</span><span class="st">'n'</span>)</a>
<a class="sourceLine" id="cb20-8" data-line-number="8"><span class="kw"><a href="https://rdrr.io/r/graphics/grid.html">grid</a></span>(<span class="dt">ny=</span><span class="ot">NULL</span>,<span class="dt">nx=</span><span class="ot">NA</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="svcnvplus_howto_files/figure-html/plot2-1.png" alt="Common breakpoints by sample" width="864"><p class="caption">
Common breakpoints by sample
</p>
</div>
</div>
<div id="breakpoint-burden" class="section level2">
<h2 class="hasAnchor">
<a href="#breakpoint-burden" class="anchor"></a>Breakpoint burden</h2>
<p>In addition to percent genome changes, we can measure the total burden of breakpoints derived from CNV segments and SV calls</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">sv_breaks_df  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sv.breaks.html">sv.breaks</a></span>(svdf)  <span class="co"># define breakpoints from SV data</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2">sv_burden &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(sv_breaks_df<span class="op">$</span>sample)</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">seg_breaks_df  &lt;-<span class="st"> </span><span class="kw"><a href="../reference/seg.breaks.html">seg.breaks</a></span>(segdf,<span class="dt">fc.pct =</span> <span class="fl">0.2</span>,<span class="dt">verbose=</span><span class="ot">FALSE</span>)  <span class="co"># define breakpoints from seg data based on certain CNV change cutoff</span></a>
<a class="sourceLine" id="cb21-4" data-line-number="4">seg_burden &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/table.html">table</a></span>(seg_breaks_df<span class="op">$</span>sample)</a>
<a class="sourceLine" id="cb21-5" data-line-number="5">common_samples &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/sets.html">intersect</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(sv_burden),<span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(seg_burden))</a>
<a class="sourceLine" id="cb21-6" data-line-number="6">dat &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(sv_burden[common_samples],seg_burden[common_samples])</a>
<a class="sourceLine" id="cb21-7" data-line-number="7"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(dat,<span class="dt">xlab=</span><span class="st">"SV burden"</span>,<span class="dt">ylab=</span><span class="st">"CNV breakpoint burden"</span>)</a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="kw"><a href="https://rdrr.io/r/graphics/legend.html">legend</a></span>(<span class="st">"topright"</span>,<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"cor="</span>,<span class="kw"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span>(dat)[<span class="dv">1</span>,<span class="dv">2</span>], <span class="dt">sep=</span><span class="st">""</span>))</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="svcnvplus_howto_files/figure-html/plot3-1.png" alt="SV versus CNV breakpoint burden" width="480"><p class="caption">
SV versus CNV breakpoint burden
</p>
</div>
</div>
<div id="identification-of-shattered-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#identification-of-shattered-regions" class="anchor"></a>Identification of shattered regions</h2>
<p>Complex chromosomal rearrangements such as chromothripsis and chromoplexy are widespread events in many cancers and may have important pathogenic roles. <code>svcnvplus</code> incorporates tools to map and visualize shattered regions across multiple samples.</p>
<p>We used LUNG cancer cell line profiles from the CCLE in order to illustrated these tools:</p>
<p>Validate segmentation and SV data.frames</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">segdf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/validate.seg.html">validate.seg</a></span>(segdat_lung_ccle)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">svdf &lt;-<span class="st"> </span><span class="kw"><a href="../reference/validate.sv.html">validate.sv</a></span>(svdat_lung_ccle)</a></code></pre></div>
<div id="chromosome-shattering-using-segmentation-data-only" class="section level3">
<h3 class="hasAnchor">
<a href="#chromosome-shattering-using-segmentation-data-only" class="anchor"></a>Chromosome shattering using segmentation data only</h3>
<p>The whole genome is binned into user defined <code>window.size</code> (Mb) and slided by <code>slide.size</code> (Mb) in order to identify regions with high CNV breakpoint density. Two cutoffs are considered for each genomic bin: * num.breaks = the minimum number of breakpoints * num.sd = the number of standard deviations above the average within a sample In addition, we evaluate the interquantile average of the distance between breakpoints in a given region: * dist.iqm.cut (default = 150000 b)</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">shatt_lung_cnv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/shattered.regions.cnv.html">shattered.regions.cnv</a></span>(segdf, <span class="dt">fc.pct =</span> <span class="fl">0.2</span>, <span class="dt">clean.brk =</span> <span class="dv">4</span>, <span class="dt">window.size =</span> <span class="dv">10</span>,</a>
<a class="sourceLine" id="cb23-2" data-line-number="2">                                        <span class="dt">slide.size =</span> <span class="dv">2</span>,<span class="dt">num.breaks =</span> <span class="dv">8</span>, <span class="dt">num.sd =</span> <span class="dv">5</span>,  </a>
<a class="sourceLine" id="cb23-3" data-line-number="3">                                        <span class="dt">dist.iqm.cut =</span> <span class="dv">150000</span>,<span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb23-4" data-line-number="4"></a>
<a class="sourceLine" id="cb23-5" data-line-number="5">shatt_lung_cnv<span class="op">$</span>regions.summary<span class="op">$</span>A549_LUNG</a></code></pre></div>
<pre><code>##   chrom     start       end nseg dist.iqm n.brk conf
## 1  chr1 110206904 112706144    5 115758.5     9   lc
## 2  chr3    377402  16241354    6 466057.7    22   HC
## 3  chr4 168970791 176576580    1 778410.0     8   HC
## 4  chr6  25347255  32530221    1 166325.0     8   HC
## 5  chr7 134278307 149847513    3 143582.3    13   lc
## 6 chr15  20586232  30894417    2 690287.2    10   HC</code></pre>
</div>
<div id="chromosome-shattering-using-segmentation-and-sv-data" class="section level3">
<h3 class="hasAnchor">
<a href="#chromosome-shattering-using-segmentation-and-sv-data" class="anchor"></a>Chromosome shattering using segmentation and SV data</h3>
<p>Analogously we can combine CNVs and SVs breakpoins to obtain a more robust evaluation of chromosome shattering (see <code><a href="../reference/shattered.regions.html">?shattered.regions</a></code>) In addition SVs provide linkage for each SV breakpoint pair which allow for an additional parametter: * interleaved.cut the minimun percentage (0-1) of interleaved SVs</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">shatt_lung &lt;-<span class="st"> </span><span class="kw"><a href="../reference/shattered.regions.html">shattered.regions</a></span>(segdf, svdf, <span class="dt">fc.pct =</span> <span class="fl">0.2</span>,  <span class="dt">min.num.probes =</span> <span class="dv">5</span>, <span class="dt">clean.brk =</span> <span class="dv">8</span>,</a>
<a class="sourceLine" id="cb25-2" data-line-number="2">                                <span class="dt">window.size =</span> <span class="dv">10</span>,<span class="dt">slide.size =</span> <span class="dv">2</span>, <span class="dt">num.seg.breaks =</span> <span class="dv">6</span>, </a>
<a class="sourceLine" id="cb25-3" data-line-number="3">                                <span class="dt">num.seg.sd =</span> <span class="dv">5</span>, <span class="dt">num.sv.breaks =</span> <span class="dv">6</span>, <span class="dt">num.sv.sd =</span> <span class="dv">5</span>, </a>
<a class="sourceLine" id="cb25-4" data-line-number="4">                                <span class="dt">num.common.breaks =</span> <span class="dv">3</span>, <span class="dt">num.common.sd =</span> <span class="dv">0</span>, <span class="dt">interleaved.cut =</span> <span class="fl">0.33</span>,</a>
<a class="sourceLine" id="cb25-5" data-line-number="5">                                <span class="dt">dist.iqm.cut =</span> <span class="dv">100000</span>,<span class="dt">verbose=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb25-6" data-line-number="6">shatt_lung<span class="op">$</span>regions.summary<span class="op">$</span>SCLC21H_LUNG</a></code></pre></div>
<pre><code>##   chrom   start       end nseg links reg.size dist.iqm.seg dist.iqm.sv
## 1  chr1 1074535   3611743    1     2 1.00e+07      58057.0    16821.67
## 2  chr8  601012 141936592   70     1 1.48e+08     151575.8   105171.70
##   n.brk.seg n.brk.sv n.orth.seg n.orth.sv interleaved conf
## 1         6       11          4         6  0.20000000   HC
## 2       567      668        446       489  0.02873563   HC</code></pre>
<p>Circos plotting is available via <a href="https://cran.r-project.org/web/packages/circlize/index.html">circlize</a> package wrapper function:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw"><a href="../reference/circ.chromo.plot.html">circ.chromo.plot</a></span>(shatt_lung,<span class="dt">sample.id =</span> <span class="st">"SCLC21H_LUNG"</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="svcnvplus_howto_files/figure-html/plot4-1.png" alt="Circos plot representing c LUNG cancer cell line with chromothripsis" width="480"><p class="caption">
Circos plot representing c LUNG cancer cell line with chromothripsis
</p>
</div>
</div>
<div id="recurrently-shattered-regions" class="section level3">
<h3 class="hasAnchor">
<a href="#recurrently-shattered-regions" class="anchor"></a>Recurrently shattered regions</h3>
<p>We evaluate the null hipothesis that shattered regions (genomic bins) may occour throughout the genome at random; To this end we first create an empirical null distribution based on the sample set we are studying. And compare with the observed distribution. This also allows to define an empirical FDR cutoff.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1">fdr.test &lt;-<span class="st"> </span><span class="kw"><a href="../reference/freq.p.test.html">freq.p.test</a></span>(shatt_lung_cnv<span class="op">$</span>high.density.regions.hc, <span class="dt">method=</span><span class="st">"bonferroni"</span>, <span class="dt">p.cut =</span> <span class="fl">0.05</span>)</a></code></pre></div>
</div>
<div id="recurrently-shattered-regions-plot" class="section level3">
<h3 class="hasAnchor">
<a href="#recurrently-shattered-regions-plot" class="anchor"></a>Recurrently shattered regions plot</h3>
<p>We can visualize the aggregate map of shattered regions for all samples with <code>recurrent.regions.plot</code></p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw"><a href="../reference/recurrent.regions.plot.html">recurrent.regions.plot</a></span>(shatt_lung_cnv, <span class="dt">fdr =</span> fdr.test<span class="op">$</span>freq.cut)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="svcnvplus_howto_files/figure-html/plot5-1.png" alt="Recurrently shattered regions map" width="864"><p class="caption">
Recurrently shattered regions map
</p>
</div>
<p>And finally collect groups of samples with shattered chromosomes in recurrent regions as defined by a given fdr cutoff</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co"># obtain genomic bins within above the FDR cutoff</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2">freq.matrix &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(shatt_lung_cnv<span class="op">$</span>high.density.regions.hc,<span class="dv">2</span>,sum)</a>
<a class="sourceLine" id="cb30-3" data-line-number="3">textRegions &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(freq.matrix <span class="op">&gt;=</span><span class="st"> </span>fdr.test<span class="op">$</span>freq.cut))</a>
<a class="sourceLine" id="cb30-4" data-line-number="4">hitRegions &lt;-<span class="st"> </span><span class="kw">remove.factors</span>((<span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(rbind,<span class="kw"><a href="https://rdrr.io/r/base/strsplit.html">strsplit</a></span>(textRegions,<span class="st">" "</span>)))))</a>
<a class="sourceLine" id="cb30-5" data-line-number="5">hitRegions[,<span class="dv">2</span>] &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(hitRegions[,<span class="dv">2</span>])</a>
<a class="sourceLine" id="cb30-6" data-line-number="6">hitRegions[,<span class="dv">3</span>] &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span>(hitRegions[,<span class="dv">3</span>])</a>
<a class="sourceLine" id="cb30-7" data-line-number="7"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(hitRegions) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"chr"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>)</a>
<a class="sourceLine" id="cb30-8" data-line-number="8"><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(hitRegions) &lt;-textRegions</a>
<a class="sourceLine" id="cb30-9" data-line-number="9"></a>
<a class="sourceLine" id="cb30-10" data-line-number="10"><span class="co"># collapes contiguous bins into unique regions</span></a>
<a class="sourceLine" id="cb30-11" data-line-number="11">bins2remove &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>()</a>
<a class="sourceLine" id="cb30-12" data-line-number="12"><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">2</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(hitRegions)){ </a>
<a class="sourceLine" id="cb30-13" data-line-number="13">  <span class="cf">if</span>(hitRegions[i,<span class="st">"chr"</span>] <span class="op">==</span><span class="st"> </span>hitRegions[i<span class="dv">-1</span>,<span class="st">"chr"</span>] ){</a>
<a class="sourceLine" id="cb30-14" data-line-number="14">    <span class="cf">if</span>(hitRegions[i,<span class="st">"start"</span>] <span class="op">&lt;</span><span class="st"> </span>(hitRegions[i<span class="dv">-1</span>,<span class="st">"end"</span>])){</a>
<a class="sourceLine" id="cb30-15" data-line-number="15">      hitRegions[i,<span class="st">"start"</span>] &lt;-<span class="st"> </span>hitRegions[i<span class="dv">-1</span>,<span class="st">"start"</span>]</a>
<a class="sourceLine" id="cb30-16" data-line-number="16">      bins2remove &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(bins2remove,textRegions[i<span class="dv">-1</span>])</a>
<a class="sourceLine" id="cb30-17" data-line-number="17">    }</a>
<a class="sourceLine" id="cb30-18" data-line-number="18">  }</a>
<a class="sourceLine" id="cb30-19" data-line-number="19">}</a>
<a class="sourceLine" id="cb30-20" data-line-number="20">hitRegionsPost&lt;-<span class="st"> </span>hitRegions[<span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/setops-methods.html">setdiff</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(hitRegions),bins2remove),]</a>
<a class="sourceLine" id="cb30-21" data-line-number="21"></a>
<a class="sourceLine" id="cb30-22" data-line-number="22"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(GenomicRanges)</a>
<a class="sourceLine" id="cb30-23" data-line-number="23">hitRegions_gr &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(hitRegions, <span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span>(chr, <span class="kw">IRanges</span>(<span class="dt">start=</span>start, <span class="dt">end=</span>end)))</a>
<a class="sourceLine" id="cb30-24" data-line-number="24">hitRegionsPost_gr &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/with.html">with</a></span>(hitRegionsPost, <span class="kw"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span>(chr, <span class="kw">IRanges</span>(<span class="dt">start=</span>start, <span class="dt">end=</span>end)))</a>
<a class="sourceLine" id="cb30-25" data-line-number="25">hits &lt;-GenomicAlignments<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/GenomicAlignments/man/findOverlaps-methods.html">findOverlaps</a></span>(hitRegionsPost_gr,hitRegions_gr)</a>
<a class="sourceLine" id="cb30-26" data-line-number="26"></a>
<a class="sourceLine" id="cb30-27" data-line-number="27">regList &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>()</a>
<a class="sourceLine" id="cb30-28" data-line-number="28"><span class="cf">for</span>(i <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/unique.html">unique</a></span>(<span class="kw">queryHits</span>(hits))) regList[[<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(hitRegionsPost[i,],<span class="dt">collapse=</span><span class="st">" "</span>) ]] &lt;-<span class="st"> </span>textRegions[<span class="kw">subjectHits</span>(hits)[<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw">queryHits</span>(hits) <span class="op">==</span><span class="st"> </span>i)]]</a>
<a class="sourceLine" id="cb30-29" data-line-number="29"><span class="co"># obtain the genomic bins with maximum number of samples</span></a>
<a class="sourceLine" id="cb30-30" data-line-number="30">regListPeak &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(regList, <span class="cf">function</span>(x) <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(freq.matrix[x] <span class="op">==</span><span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(freq.matrix[x]))))</a>
<a class="sourceLine" id="cb30-31" data-line-number="31"><span class="co"># collect samples with shattered region in the peaks </span></a>
<a class="sourceLine" id="cb30-32" data-line-number="32">regListPeakSamples &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(regListPeak, <span class="cf">function</span>(x) <span class="kw"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/apply.html">apply</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(shatt_lung_cnv<span class="op">$</span>high.density.regions.hc[,x]),<span class="dv">1</span>,sum) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)))</a></code></pre></div>
<p>Beyond this point the user can test case/control hipothesys for chromosome shattering of specific genomic regionsâ€¦</p>
</div>
</div>
<div id="recurrently-altered-genes" class="section level2">
<h2 class="hasAnchor">
<a href="#recurrently-altered-genes" class="anchor"></a>Recurrently altered genes</h2>
<div id="amplifications-and-deep-deletions" class="section level3">
<h3 class="hasAnchor">
<a href="#amplifications-and-deep-deletions" class="anchor"></a>amplifications and deep deletions</h3>
</div>
<div id="recurrently-altered-genes-using-cnv-breakpoints" class="section level3">
<h3 class="hasAnchor">
<a href="#recurrently-altered-genes-using-cnv-breakpoints" class="anchor"></a>Recurrently altered genes using CNV breakpoints</h3>
<p>Instead of focusing on high-level dosage changes, we evaluate whether CNV break ends overlap with known genes or upstream regions.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1">results_cnv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cnv.gene.breaks.html">cnv.gene.breaks</a></span>(nbl_segdat, <span class="dt">fc.pct =</span> <span class="fl">0.2</span>, <span class="dt">genome.v=</span><span class="st">"hg19"</span>,<span class="dt">clean.brk =</span> <span class="dv">8</span>,<span class="dt">upstr =</span> <span class="dv">50000</span>,<span class="dt">verbose=</span><span class="ot">FALSE</span>)</a></code></pre></div>
<p>We can obtain a ranking of altered genes by breakpoints overlapping their transcriptions start and endâ€¦</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(results_cnv<span class="op">$</span>geneSamples,length)),<span class="dt">decreasing=</span>T)[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>],<span class="dt">las=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="svcnvplus_howto_files/figure-html/plot6-1.png" alt="Recurrently altered genes with overlapping CNV breakpoints" width="864"><p class="caption">
Recurrently altered genes with overlapping CNV breakpoints
</p>
</div>
<p>â€¦ or upstream (50Kb)</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(results_cnv<span class="op">$</span>upstreamSamples,length)),<span class="dt">decreasing=</span>T)[<span class="dv">1</span><span class="op">:</span><span class="dv">20</span>],<span class="dt">las=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure" style="text-align: center">
<img src="svcnvplus_howto_files/figure-html/plot7-1.png" alt="Recurrently altered genes with overlapping CNV breakpoints" width="864"><p class="caption">
Recurrently altered genes with overlapping CNV breakpoints
</p>
</div>
</div>
<div id="recurrently-altered-genes-using-sv-breakpoints" class="section level3">
<h3 class="hasAnchor">
<a href="#recurrently-altered-genes-using-sv-breakpoints" class="anchor"></a>Recurrently altered genes using SV breakpoints</h3>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#index">Index:</a></li>
      <li><a href="#install-svcnvplus">Install svcnvplus</a></li>
      <li><a href="#input-data">Input data</a></li>
      <li><a href="#validate-data-types">Validate data types</a></li>
      <li><a href="#cnv-analysys-and-visualization">CNV analysys and visualization</a></li>
      <li><a href="#co-localization-of-breakpoints">Co-localization of breakpoints</a></li>
      <li><a href="#breakpoint-burden">Breakpoint burden</a></li>
      <li><a href="#identification-of-shattered-regions">Identification of shattered regions</a></li>
      <li><a href="#recurrently-altered-genes">Recurrently altered genes</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gonzalo Lopez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
