<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ svcnvplus</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">svcnvplus</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/svcnvplus_howto.html">svcnvplus_howto</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="IQM.html">IQM()</a></code> </p>
        </td>
        <td><p>Obtains inter quantile average for a defined 'x' vector and both lower and upper quantiles</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="IQSD.html">IQSD()</a></code> </p>
        </td>
        <td><p>Obtains inter quantile standard deviation for a defined 'x' vector and both lower and upper quantiles</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ave.segmean.html">ave.segmean()</a></code> </p>
        </td>
        <td><p>Obtain the average weighted segment mean from a segmentaton file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="break.density.html">break.density()</a></code> </p>
        </td>
        <td><p>Identify CNV breakpoints provided a segmentation file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chr.arm.cnv.html">chr.arm.cnv()</a></code> </p>
        </td>
        <td><p>Obtain a matrix with the weighted average CN per chromosome arm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chromosome.limit.coords.html">chromosome.limit.coords()</a></code> </p>
        </td>
        <td><p>This function mapps defines mapping limmit regions based on a segmentation file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circ.chromo.plot.html">circ.chromo.plot()</a></code> </p>
        </td>
        <td><p>Circos plot combining segmentation and SV calls</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="circ.wg.plot.html">circ.wg.plot()</a></code> </p>
        </td>
        <td><p>Circos plot combining segmentation and SV calls</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="clean.cnv.artifact.html">clean.cnv.artifact()</a></code> </p>
        </td>
        <td><p>Obtain a matrix with the weighted average CN per chromosome arm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cnv.freq.plot.html">cnv.freq.plot()</a></code> </p>
        </td>
        <td><p>Plot CNV frequency across the human genome from a seg file sontaining multiple samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="cnv.gene.breaks.html">cnv.gene.breaks()</a></code> </p>
        </td>
        <td><p>Identify recurrently altered genes</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createRandomString.html">createRandomString()</a></code> </p>
        </td>
        <td><p>Obtains inter quantile standard deviation for a defined 'x' vector and both lower and upper quantiles</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="freq.p.test.html">freq.p.test()</a></code> </p>
        </td>
        <td><p>Obtains significance cutoff for the frequency of binary events encoded in a matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="gene.cnv.html">gene.cnv()</a></code> </p>
        </td>
        <td><p>Obtain a matrix with the weighted average CN per chromosome arm</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="match.breaks.html">match.breaks()</a></code> </p>
        </td>
        <td><p>Identify CNV breakpoints provided a segmentation file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="match.variant.breaks.html">match.variant.breaks()</a></code> </p>
        </td>
        <td><p>Identify CNV breakpoints provided a segmentation file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="med.segmean.html">med.segmean()</a></code> </p>
        </td>
        <td><p>Obtain the median segment mean from a segmentaton file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="merge2lists.html">merge2lists()</a></code> </p>
        </td>
        <td><p>Given two lists with (or without common names) returns a combined list; if common names their values are returned unique</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pct.genome.changed.html">pct.genome.changed()</a></code> </p>
        </td>
        <td><p>Calculate the percentage of genome changed by CNV</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="recurrent.regions.plot.html">recurrent.regions.plot()</a></code> </p>
        </td>
        <td><p>Plot CNV frequency across the human genome from a seg file sontaining multiple samples</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="seg.breaks.html">seg.breaks()</a></code> </p>
        </td>
        <td><p>Identify read-depth breakpoints provided a segmentation file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="segdat_lung_ccle.html">segdat_lung_ccle</a></code> </p>
        </td>
        <td><p>CCLE degmentation data from breast tissue cell lines</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="segment.gap.html">segment.gap()</a></code> </p>
        </td>
        <td><p>Filling gaps in a segmentation dat abased on chromosome limits defined withing the complete segmentation dataset</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="shattered.eval.html">shattered.eval()</a></code> </p>
        </td>
        <td><p>Evaluate shattered regions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="shattered.regions.html">shattered.regions()</a></code> </p>
        </td>
        <td><p>Caller for shattered genomic regions based on breakpoint densities</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="shattered.regions.cnv.html">shattered.regions.cnv()</a></code> </p>
        </td>
        <td><p>Caller for shattered genomic regions based on breakpoint densities</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sv.breaks.html">sv.breaks()</a></code> </p>
        </td>
        <td><p>Identify SV call breakpoints and filter low cov regions if provided</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="validate.seg.html">validate.seg()</a></code> </p>
        </td>
        <td><p>This function calculates the percent genome changed by CNV</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="validate.sv.html">validate.sv()</a></code> </p>
        </td>
        <td><p>This function calculates the percent genome changed by CNV
sv classes accepted: DEL(deletion), DUP(duplication), h2hINV(head to head inversion), t2tINV(tail to tail inversion), TRA(translocation), INV(inversion)</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Gonzalo Lopez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


